common/src/main/java/feast/common/logging/LogEvent.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2018-2019 The Feast Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.common.logging;
+
+import com.google.auto.value.AutoOneOf;
+import com.google.auto.value.AutoValue;
+import com.google.protobuf.Message;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+/** LogEvent describes a specific event that a log is capturing */
+@AutoOneOf(LogEvent.Kind.class)
+public abstract class LogEvent {
```

Comment:
My initial reaction when seeing this code is that it seems quite overwhelming. A lot of decisions have been made here in terms of the structure of our audit logging and I am not sure if we are being too ambitious here. The amount of code that needs to be maintained here is quite intense, and I am quite worried that these abstractions will end up changing quite soon.

Did you follow some kind of convention when structuring the audit logger and its associated types, or was everything custom made?

Reply:
> Did you follow some kind of convention when structuring the audit logger and its associated types, or was everything custom made?

No, Yes.


> I was hoping we could start with a simpler key/value API.

Reasons that we should go with a Key Value API.
- Infinitely more flexible and extensible as no constraints are applied.

Reasons that we should not go with a Key Value API:
- Java is statically typed. Going with a generic Key Value would mean `Map<String, Object>` which implies lots of ugly casting.
- With no enforced structure in place it will be difficult to parse the logs in upstream systems as there is no expected structure.
- Enforces some rigidity to logging so that we don't break upstream systems unknowing built upon audit logging.

The structured log will be depended on by upstream logging systems and should be treated as a user facing API that should be clearly defined, the same way Core and Serving Service API are clearly defined.

>  I am quite worried that these abstractions will end up changing quite soon once people start trying to use the audit logger

The important thing is that we are fully aware of changes instead of being blind sighted by a `map.put()`

Reply:
My suggestion is not to throw away all static typing and just accept arbitrary string/object keys and values. Parts of what you have implemented makes sense, but other parts do not.

From a functional perspective what we want is for specific parts of the Feast code base to create a message that can be serialized and printed/sent to the audit log. so having static methods like `ofMessage(MessageLogEvent.Kind kind, String method, Message message, String identity)` makes sense. 

However, I question the value of this `LogEvent` as a container. It seems like a catchall class that provides little in terms of a contract. When would you have both a `MessageLogEvent` and `ActionLogEvent` in the same object? It doesnt seem like that should be possible and so its not clear what the scope of this `LogEvent` class should be. Will it just grow infinitely as a catchall class that can contain any subclass?

If we agree that maintaining state is a bad thing, then do we need to maintain LogEvents as an object? What is the SLF4J contract that we need to meet. Does it just expect strings?

What prevents us from having an `AuditLogger` class with static methods like `ofMessage` that converts a tight contract of objects into a Map or string that can be serialized?

Reply:
Refactored away LogEvent into MessageAuditLogEntry, ActionAuditLogEntry, TransitionAuditLogEntry subclasses.